### PR TITLE
feat: produce updater on request

### PR DESCRIPTION
**Pull request by mistydemeo**
This is the first part of installing the updater for those who plan to use it. In this phase, a new setting allows users to opt into creating the installer and adding it to the artifacts list; if requested, it'll be added to the GitHub artifacts along with the software itself, with one copy per target triple.

In the next phase, the shell and PowerShell installers will be updated to check for and install the updater if present, renaming it to the appropriate `appname-updater` filename.

Note that right now this *only* produces the updater via `cargo install`ing from `HEAD` in the repo, but once we have the first stable release there will be prebuilt binaries in the axoupdater repo; in that case, we can simply pull down the latest binary for the target triple and only fall back to building from source if there's no binary for the requested triple.

### CODE REVIEW

File: cargo-dist/src/config.rs
Comment (Gankra) [inline cargo-dist/src/config.rs:514]: *thinks* this seems like it should be an app-scoped config, yeah? since each app gets its own installer-sh (can ship it as top-level, but, it *can* be app-scoped, at least).

File: cargo-dist/src/tasks.rs
Comment (Gankra) [inline cargo-dist/src/tasks.rs:1425]: can't this just be `{variant.id}-{update}`?

Comment (Gankra) [reply cargo-dist/src/tasks.rs:1425]: also is there a reason we need to generate N copies of this? We could just make one universal/shared one and the installer could just rename it on fetch..?

(tbh i expected the design here to be putting it inside the archives... hmm... interesting)

File: cargo-dist/src/tasks.rs
Comment (Gankra) [inline cargo-dist/src/tasks.rs:1707]: once again variant.id is for this

Comment (Gankra) [reply cargo-dist/src/tasks.rs:1707]: actually you just shouldn't be computing this twice. one of these two should set the other.

File: cargo-dist/src/tasks.rs
Comment (Gankra) [inline cargo-dist/src/tasks.rs:2051]: once again, centralize this

File: cargo-dist/src/tasks.rs
Comment (mistydemeo) [reply cargo-dist/src/tasks.rs:1425]: Keeping it separate makes it easier to avoid fetching and installing it in the Homebrew and npm installers; that way there's no risk of it leaking into inappropriate places.

> We could just make one universal/shared one and the installer could just rename it on fetch..?

The updater is native code; we need one per target. We choose which version to fetch and install in the installers.

File: cargo-dist/src/tasks.rs
Comment (mistydemeo) [reply cargo-dist/src/tasks.rs]: Fixed; swapped to something similar to `make_executable_zip_for_variant`.

File: cargo-dist/src/config.rs
Comment (mistydemeo) [reply cargo-dist/src/config.rs:518]: Oh yes, good call. Changed.

**Comment by mistydemeo** [comment]
Made the changes you suggested and pushed. Tests are passing, and I've tested this with my cargodisttest repo using the shell installer; it's installing the updater, and it runs fine. I think we're about ready to merge this, assuming we're ready to cut an initial release of the updater itself.

**Comment by mistydemeo** [comment]
Tested actually running the update, and it's working fine:

```
$ axolotlsay-update

Checking for updates...
downloading axolotlsay 0.2.106 x86_64-apple-darwin
installing to /Users/mistydemeo/.cargo/bin
  axolotlsay
  axolotlsay-update
everything's installed!
New release installed!
```

File: cargo-dist/src/lib.rs
Comment (Gankra) [inline cargo-dist/src/lib.rs:176]: hmm in an ideal world we can fetch the latest release's prebuilt binaries? (with fallback to this for unbuilt platforms)

File: cargo-dist/src/lib.rs
Comment (Gankra) [inline cargo-dist/src/lib.rs:184]: i'm fairly certain we have this kind of logic somewhere else, and should reuse/factor it.

File: cargo-dist/src/tasks.rs
Comment (Gankra) [inline cargo-dist/src/tasks.rs:1432]: might be worth adding a comment here being like "yep this makes an updater for every app, it could be shared between apps, but for an mvp it's fine".

File: cargo-dist/src/tasks.rs
Comment (Gankra) [inline cargo-dist/src/tasks.rs:1760]: lolsob (it's fine, just, oh no my crimes)

File: cargo-dist/templates/installer/installer.ps1.j2
Comment (Gankra) [inline cargo-dist/templates/installer/installer.ps1.j2:81]: ...i suppose that is the easiest way. makes me wonder if this should just be in the "artifacts" input above though

File: cargo-dist/templates/installer/installer.ps1.j2
Comment (Gankra) [inline cargo-dist/templates/installer/installer.ps1.j2:191]: do we want to add printouts for "unpacked updater"?

**Review by Gankra — APPROVED** [review]
Looks reasonable...

File: cargo-dist/src/lib.rs
Comment (mistydemeo) [reply cargo-dist/src/lib.rs:176]: Yes! Right now it's not done just because those binaries don't exist yet. This is the fallback until that happens. Likewise, it'll be built from the crates.io release instead of git.

Comment (ashleygwilliams) [reply cargo-dist/src/lib.rs:176]: ignore me if i'm way off here, but i'm curious why we'd prefer the crates.io release over the git one (i think we often can't be sure there is a crates.io release, whereas we always know there is a git one)

Comment (mistydemeo) [reply cargo-dist/src/lib.rs:176]: This is of axoupdater, rather than the user's own software. Because we publish axoupdater ourselves, we can know that we'll always have a stable version on crates.io.

File: cargo-dist/src/lib.rs
Comment (mistydemeo) [reply cargo-dist/src/lib.rs:184]: There is another place, but it operates on strings of base filenames: https://github.com/axodotdev/cargo-dist/blob/main/cargo-dist/src/tasks.rs#L1150-L1151

File: cargo-dist/templates/installer/installer.ps1.j2
Comment (mistydemeo) [reply cargo-dist/templates/installer/installer.ps1.j2:191]: The list of binaries will already include a thing that ends in `-update`, but I could add an extra message on top of that if you think it'd help legibility.

File: cargo-dist/templates/installer/installer.ps1.j2
Comment (mistydemeo) [reply cargo-dist/templates/installer/installer.ps1.j2:81]: Yeah, I wondered about that. In practice doing that had turned out to be a bit less convenient than I'd like, so this was easier for now.

File: cargo-dist/Cargo.toml
Comment (ashleygwilliams) [inline cargo-dist/Cargo.toml:39]: i assume we won't merge this until we publish the axo asset version with this feature?

File: cargo-dist/Cargo.toml
Comment (ashleygwilliams) [inline cargo-dist/Cargo.toml]: honestly love this

**Review by ashleygwilliams — COMMENTED** [review]
just looking at the last commit all i can say is i love moving logic into axoasset haha, really cleans things up!

File: cargo-dist/Cargo.toml
Comment (mistydemeo) [reply cargo-dist/Cargo.toml]: Made that release and swapped over to it!

**Pull request merged**
